From: =?utf-8?q?Lo=C3=AFc_Minier?= <lool@dooz.org>
Date: Tue, 15 Aug 2006 09:54:12 +0000
Subject: Don't build perf measurement tools in static builds

Bug: http://bugzilla.gnome.org/show_bug.cgi?id=346559
---
 Makefile.am | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/Makefile.am b/Makefile.am
index 8e3a2f1..1fdcbc4 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,7 +1,12 @@
 ## Makefile.am for GTK+
 include $(top_srcdir)/Makefile.decl
 
-SRC_SUBDIRS = gdk gtk modules demos tests perf
+# don't build perf measurement tools in static builds
+if ENABLE_SHARED
+MAYBE_PERF = perf
+endif
+
+SRC_SUBDIRS = gdk gtk modules demos tests $(MAYBE_PERF)
 SUBDIRS = po po-properties $(SRC_SUBDIRS) docs m4macros build
 
 # require automake 1.4
