Index: gtk+2.0-2.21.8/gtk/gtktreeview.c
===================================================================
--- gtk+2.0-2.21.8.orig/gtk/gtktreeview.c	2010-09-10 15:07:08.000000000 +0200
+++ gtk+2.0-2.21.8/gtk/gtktreeview.c	2010-09-14 20:23:25.116679103 +0200
@@ -9623,7 +9623,8 @@
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
