Index: gtk+2.0-2.22.0/gtk/gtktreeview.c
===================================================================
--- gtk+2.0-2.22.0.orig/gtk/gtktreeview.c	2010-09-23 05:00:37.000000000 +0200
+++ gtk+2.0-2.22.0/gtk/gtktreeview.c	2010-09-23 20:29:38.869590760 +0200
@@ -9623,7 +9623,8 @@
 
       if (cursor_path)
 	{
-	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE)
+	  if (tree_view->priv->selection->type == GTK_SELECTION_MULTIPLE ||
+	      tree_view->priv->selection->type == GTK_SELECTION_SINGLE)
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, FALSE, FALSE);
 	  else
 	    gtk_tree_view_real_set_cursor (tree_view, cursor_path, TRUE, FALSE);
