Index: gtk+2.0-2.21.8/gtk/updateiconcache.c
===================================================================
--- gtk+2.0-2.21.8.orig/gtk/updateiconcache.c	2010-09-10 15:06:45.000000000 +0200
+++ gtk+2.0-2.21.8/gtk/updateiconcache.c	2010-09-14 20:23:28.581689463 +0200
@@ -660,7 +660,7 @@
 		      directories = g_list_append (directories, g_strdup (subdir));
 		    }
 		  else
-		    dir_index = 0xffff;
+		    continue;
 		}
 		
 	      image = g_new0 (Image, 1);
