Index: gtk+2.0-2.21.8/gdk/x11/gdkdnd-x11.c
===================================================================
--- gtk+2.0-2.21.8.orig/gdk/x11/gdkdnd-x11.c	2010-09-15 11:31:48.162871918 +0200
+++ gtk+2.0-2.21.8/gdk/x11/gdkdnd-x11.c	2010-09-15 11:31:50.781884886 +0200
@@ -570,6 +570,7 @@
 
   g_free (children);
 
+#ifdef HAVE_XCOMPOSITE
   /*
    * Add the composite overlay window to the cache, as this can be a reasonable
    * Xdnd proxy as well.
@@ -583,6 +584,7 @@
       gdk_window_cache_add (result, cow, 0, 0, gdk_screen_get_width (screen), gdk_screen_get_height (screen), TRUE);
       XCompositeReleaseOverlayWindow (xdisplay, GDK_WINDOW_XWINDOW (root_window));
     }
+#endif
 
   return result;
 }
